File: backend-java/src/main/java/com/bffgin/backend/auth/HmacVerifier.java
package com.bffgin.backend.auth;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.security.Keys;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;

/** ローカルHMAC発行(iss=Dispatcher.LOCAL_HMAC_ISSUER)の検証。HS256、共有シークレット */
public final class HmacVerifier implements TokenVerifier {

    private final SecretKey key;
    private final String issuer;
    private final String audience;

    public HmacVerifier(String secret, String issuer, String audience) {
        this.key = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
        this.issuer = issuer;
        this.audience = audience;
    }

    @Override
    public Claims verify(String token) throws VerifyException {
        try {
            var jws = Jwts.parser()
                    .verifyWith(key)
                    .build()
                    .parseSignedClaims(token);

            // 【セキュリティ上の確認、C/C++実装と同じ観点】jjwtはverifyWith()に渡した鍵の型
            // (SecretKey=HMAC系)と整合しないアルゴリズム(RS256等)のトークンをそもそも受理しないが、
            // ヘッダのalgが期待通りHS256であることも明示的に確認する(多層防御)
            String alg = jws.getHeader().getAlgorithm();
            if (!"HS256".equals(alg)) {
                throw new VerifyException("unexpected alg: " + alg);
            }

            io.jsonwebtoken.Claims claims = jws.getPayload();
            if (!issuer.equals(claims.getIssuer())) {
                throw new VerifyException("issuer mismatch");
            }
            if (!AudienceMatcher.matches(claims, audience)) {
                throw new VerifyException("audience mismatch");
            }
            String azp = claims.get("azp", String.class);
            return new Claims(claims.getSubject(), claims.getIssuer(), azp == null ? "" : azp);
        } catch (JwtException e) {
            throw new VerifyException("HMAC verify failed: " + e.getMessage(), e);
        }
    }
}

File: backend-kotlin/src/main/kotlin/com/bffgin/backend/auth/HmacVerifier.kt
package com.bffgin.backend.auth

import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import java.nio.charset.StandardCharsets
import javax.crypto.SecretKey

/** ローカルHMAC発行(iss=Dispatcher.LOCAL_HMAC_ISSUER)の検証。HS256、共有シークレット */
class HmacVerifier(secret: String, private val issuer: String, private val audience: String) : TokenVerifier {

    private val key: SecretKey = Keys.hmacShaKeyFor(secret.toByteArray(StandardCharsets.UTF_8))

    override suspend fun verify(token: String): Claims {
        try {
            val jws = Jwts.parser()
                .verifyWith(key)
                .build()
                .parseSignedClaims(token)

            // 【セキュリティ上の確認、backend-javaと同じ観点】jjwtはverifyWith()に渡した鍵の型
            // (SecretKey=HMAC系)と整合しないアルゴリズム(RS256等)のトークンをそもそも受理しないが、
            // ヘッダのalgが期待通りHS256であることも明示的に確認する(多層防御)
            val alg = jws.header.algorithm
            if (alg != "HS256") {
                throw VerifyException("unexpected alg: $alg")
            }

            val claims = jws.payload
            if (claims.issuer != issuer) {
                throw VerifyException("issuer mismatch")
            }
            if (!AudienceMatcher.matches(claims, audience)) {
                throw VerifyException("audience mismatch")
            }
            return Claims(claims.subject, claims.issuer, claims.get("azp", String::class.java) ?: "")
        } catch (e: VerifyException) {
            throw e
        } catch (e: JwtException) {
            throw VerifyException("HMAC verify failed: ${e.message}", e)
        }
    }
}
